File: solutions/java/src/main/java/com/epi/ValidParentheses.java
package com.epi;

import java.util.LinkedList;

public class ValidParentheses {
  // @include
  public static boolean isWellFormed(String s) {
    LinkedList<Character> leftChars = new LinkedList<>();
    for (int i = 0; i < s.length(); ++i) {
      if (s.charAt(i) == '(' || s.charAt(i) == '{' || s.charAt(i) == '[') {
        leftChars.push(s.charAt(i));
      } else {
        if (leftChars.isEmpty()) {
          return false; // Unmatched right char.
        }
        if ((s.charAt(i) == ')' && leftChars.peek() != '(') ||
            (s.charAt(i) == '}' && leftChars.peek() != '{') ||
            (s.charAt(i) == ']' && leftChars.peek() != '[')) {
          return false; // Mismatched chars.
        }
        leftChars.pop();
      }
    }
    return leftChars.isEmpty();
  }
  // @exclude

  private static void smallTest() {
    assert(isWellFormed("()"));
    assert(isWellFormed("()[]{}"));
    assert(isWellFormed("[()[]]{}"));
    assert(isWellFormed("(()[]{()[]{}{}})"));
    assert(!isWellFormed("([)]"));
    assert(!isWellFormed("["));
    assert(!isWellFormed("(()[]{()[]{})({}})"));
  }

  public static void main(String[] args) {
    smallTest();
    if (args.length == 1) {
      System.out.println(isWellFormed(args[0]));
    }
  }
}

File: static/solutions/cpp/valid-parentheses.cc
// Copyright (c) 2013 Elements of Programming Interviews. All rights reserved.

#include <cassert>
#include <iostream>
#include <stack>
#include <string>

using std::boolalpha;
using std::cout;
using std::endl;
using std::stack;
using std::string;

// @include
bool IsWellFormed(const string& s) {
  stack<char> left_chars;
  for (int i = 0; i < s.size(); ++i) {
    if (s[i] == '(' || s[i] == '{' || s[i] == '[') {
      left_chars.emplace(s[i]);
    } else {
      if (left_chars.empty()) {
        return false;  // Unmatched right char.
      }
      if ((s[i] == ')' && left_chars.top() != '(') ||
          (s[i] == '}' && left_chars.top() != '{') ||
          (s[i] == ']' && left_chars.top() != '[')) {
        return false;  // Mismatched chars.
      }
      left_chars.pop();
    }
  }
  return left_chars.empty();
}
// @exclude

void small_test() {
  assert(IsWellFormed("()"));
  assert(IsWellFormed("()[]{}"));
  assert(IsWellFormed("[()[]]{}"));
  assert(IsWellFormed("(()[]{()[]{}{}})"));
  assert(!IsWellFormed("([)]"));
  assert(!IsWellFormed("["));
  assert(!IsWellFormed("(()[]{()[]{})({}})"));
}

int main(int argc, char** argv) {
  small_test();
  if (argc == 2) {
    cout << boolalpha << IsWellFormed(argv[1]) << endl;
  }
  return 0;
}
